dataframe/series.go

```diff
@@ -0,0 +1,423 @@
+	var tmpl string
+	if len(s.index) == 0 {
+		// Result looks like '%-2d    %6s'
+		tmpl = fmt.Sprintf("%%-%dd%s%%%ds", indexWidth, padding, colWidth)
+	} else {
+		// Result looks like '%-4s    %6s'
+		tmpl = fmt.Sprintf("%%-%ds%s%%%ds", indexWidth, padding, colWidth)
+	}
+
+	// Render each value in the series
+	render := make([]string, 0, s.len()+1)
+	for i, elem := range s.stringValues() {
+		line := ""
+		if len(s.index) == 0 {
+			line = fmt.Sprintf(tmpl, i, elem)
+		} else {
+			line = fmt.Sprintf(tmpl, s.index[i], elem)
+		}
+		render = append(render, line)
+	}
+
+	// Combine the lines together
+	render = append(render, epilogue)
+	return strings.Join(render, "\n")
+}
+
+// values returns a slice of some go native type
+func (s *Series) values() []interface{} {
+	if s.which == TypeInt {
+		result := make([]interface{}, len(s.valInts))
+		for i, elem := range s.valInts {
+			result[i] = elem
+		}
+		return result
+	} else if s.which == TypeFloat {
+		result := make([]interface{}, len(s.valFloats))
+		for i, elem := range s.valFloats {
+			result[i] = elem
+		}
+		return result
+	}
+	result := make([]interface{}, len(s.valObjs))
+	for i, elem := range s.valObjs {
+		result[i] = elem
+	}
+	return result
+}
+
+// stringValues returns a slice of the stringified values, fit for printing
+func (s *Series) stringValues() []string {
+	if s.which == TypeInt {
+		result := make([]string, len(s.valInts))
+		for i, elem := range s.valInts {
+			result[i] = strconv.Itoa(elem)
+		}
+		return result
+	} else if s.which == TypeFloat {
+		result := make([]string, len(s.valFloats))
+		for i, elem := range s.valFloats {
+			result[i] = fmt.Sprintf("%1.1f", elem)
+		}
+		return result
+	}
+	return s.valObjs
+}
+
+// len returns the number of values
+func (s *Series) len() int {
+	if s.which == TypeInt {
+		return len(s.valInts)
+	} else if s.which == TypeFloat {
+		return len(s.valFloats)
+	}
+	return len(s.valObjs)
+}
+
+// takeFirst returns a series containing just the first 'n' values
+func (s *Series) takeFirst(n int) Series {
+	var valInts []int
+	var valFloats []float64
+	var valObjs []string
+	if s.which == TypeInt {
+		valInts = s.valInts[:n]
+	} else if s.which == TypeFloat {
+		valFloats = s.valFloats[:n]
+	} else {
+		valObjs = s.valObjs[:n]
+	}
+	return Series{
+		which:     s.which,
+		valInts:   valInts,
+		valFloats: valFloats,
+		valObjs:   valObjs,
+		//index: s.index,
+		dtype: s.dtype,
+		name:  s.name,
+	}
+}
+
+// strAt returns the cell at position 'i', as a string fit for printing
+func (s *Series) strAt(i int) string {
+	if s.which == TypeInt {
+		return strconv.Itoa(s.valInts[i])
+	} else if s.which == TypeFloat {
+		return fmt.Sprintf("%1.1f", s.valFloats[i])
+	}
+	return s.valObjs[i]
+}
+
+// at returns the cell at position 'i' as a go native type
+func (s *Series) at(i int) interface{} {
+	if s.which == TypeInt {
+		return s.valInts[i]
+	} else if s.which == TypeFloat {
+		return s.valFloats[i]
+	}
+	return s.valObjs[i]
+}
+
```

Comment:
I don't think we need this for `Series`, but on the forthcoming `Dataframe` object we'll need a go-side constructor function, and _that_ constructor may need a go-side series constructor. They'll need to be exported. I think we should save actually writing the constructors for a separate PR added after we're done initial implementations.

Reply:
Removed. It'll be needed for convenience for `DataFrame`, but not currently.

Reply:
erm, sorry, I meant _adding_ a Constructor function that a go program can call without dealing with starlark values. I'm totally fine with the starlark-facing constructor sticking around